nodejs-env/src/main/scala/org/scalajs/jsenv/nodejs/NodeJSEnv.scala

```diff
@@ -9,63 +9,95 @@
 
 package org.scalajs.jsenv.nodejs
 
+import java.nio.charset.StandardCharsets
+
 import scala.collection.immutable
+import scala.collection.JavaConverters._
 
 import org.scalajs.jsenv._
-
+import org.scalajs.core.ir.Utils.escapeJS
 import org.scalajs.core.tools.io._
 import org.scalajs.core.tools.logging._
 
-import java.io.{ Console => _, _ }
-
-class NodeJSEnv(config: NodeJSEnv.Config) extends AbstractNodeJSEnv {
+import java.io._
 
+final class NodeJSEnv(config: NodeJSEnv.Config) extends JSEnv {
   def this() = this(NodeJSEnv.Config())
 
-  protected def vmName: String = "Node.js"
-
-  protected def executable: String = config.executable
-
-  override protected def args: immutable.Seq[String] = config.args
+  val name: String = "Node.js"
 
-  override protected def env: Map[String, String] = config.env
+  def start(input: Input, runConfig: RunConfig): JSRun =
+    internalStart(initFiles ++ inputFiles(input), runConfig)
 
-  // TODO Our Build wants this to be public, but it does not seem clean
-  override def wantSourceMap: Boolean = config.sourceMap
+  def startWithCom(input: Input, runConfig: RunConfig,
+      onMessage: String => Unit): JSComRun = {
+    ComRun.start(runConfig, onMessage) { comLoader =>
+      val files = initFiles ::: (comLoader :: inputFiles(input))
+      internalStart(files, runConfig)
+    }
+  }
 
-  override def jsRunner(files: Seq[VirtualJSFile]): JSRunner =
-    new NodeRunner(files)
+  private def internalStart(files: List[VirtualJSFile],
+      runConfig: RunConfig): JSRun = {
+    val command = config.executable :: config.args
+    val externalConfig = ExternalJSRun.Config()
+      .withEnv(env)
+      .withRunConfig(runConfig)
+    ExternalJSRun.start(command, externalConfig)(NodeJSEnv.write(files))
+  }
 
-  override def asyncRunner(files: Seq[VirtualJSFile]): AsyncJSRunner =
-    new AsyncNodeRunner(files)
+  private def initFiles: List[VirtualJSFile] = {
+    val base = List(Support.fixPercentConsole, NodeJSEnv.runtimeEnv)
 
-  override def comRunner(files: Seq[VirtualJSFile]): ComJSRunner =
-    new ComNodeRunner(files)
+    if (config.sourceMap) NodeJSEnv.installSourceMap :: base
+    else base
+  }
 
-  protected class NodeRunner(files: Seq[VirtualJSFile])
-      extends ExtRunner(files) with AbstractBasicNodeRunner
+  private def inputFiles(input: Input) = input match {
+    case Input.FilesToLoad(files) => files
+    case _ =>
+      throw new IllegalArgumentException(
+        s"$name does not support input $input")
+  }
 
-  protected class AsyncNodeRunner(files: Seq[VirtualJSFile])
-      extends AsyncExtRunner(files) with AbstractBasicNodeRunner
+  private def env: Map[String, String] = {
+    System.getenv().asScala.toMap + ("NODE_MODULE_CONTEXTS" -> "0") ++ config.env
```

Comment:
This should probably be part of `ExternalJSRun`. I'm just not sure how to deal with the merging of properties.

Reply:
Hah, turns out inheriting the environment is the default of process builder. So I just need to remove the line that clears it.